### ISSUE TITLE
Using Tor with socks5 leaks un-anonymized traffic in several situations

### ISSUE THREAD

**Issue by missytake** [deltachat/deltachat-desktop#3093]
Under Linux, there is the use case of using the experimental socks5 proxy support to tunnel Delta Chat traffic through the tor daemon, thereby anonymizing the IP address while using Delta Chat. (For an example use case, see this guide: https://lefherz.net/2022/08/28/how-to-setup-delta-chat-on-tails-in-4-steps/ )

Unfortunately, the socks5 support is still experimental, and was never audited. So far, it seems there are several situations in which the IP address is leaked to different parties in the network. None of them are a severe issue and can be easily used to de-anonymize the user, but in combination with other data de-anonymization could happen. The socks5 support is experimental and comes with a warning:

![Screenshot_2023-01-24_12-17-45](https://user-images.githubusercontent.com/57442838/214278309-edc7b207-7707-4a0f-870c-1f3c894e100f.png)

Still, this situation needs to be improved to protect Delta Chat users. Let's use this issue to collect cases in which the IP address can get leaked, and to fix them.

1. One of them is already described in the warning - during onboarding, there is an unencrypted DNS MX lookup whether the domain's MX entry belongs to Google Workspaces. This tells a potential network attacker and the DNS server that this IP address plans to use an email address at this certain domain. The DNS server could compile a list of IP addresses which asked for this MX entry, substract IP addresses which are obviously mail servers and just need the MX entry for relaying emails, and have a list of IP addresses using this specific mail domain with Delta Chat. For mail domains with only one or few users, and with additional information about the Delta Chat usage of the target, this can leak the IP address of the target to the attacker.
2. HTTP traffic is not tunneled through socks5 yet, as https://github.com/async-email/async-native-tls doesn't support this (yet). Delta Chat opens HTTP(S) connections in several situations:
  1. during onboarding, if the mail domain of the user is not in the provider database, the autoconfig XML is requested over HTTPS to find out server configuration. This can leak the IP address of the target to the mail server. This information on its own does not suffice to identify the user account connected to this IP address, but correlating it with SMTP/IMAP authentication logs could achieve that, depending on how many other users the email server has.
  2. when using the location tracking feature, mapbox tiles are downloaded over HTTPS to display the location of chat partners on a map. This can leak the IP address of the sender and/or recipient of the location message to mapbox.com, if they view the location map. mapbox.com could connect this information with the requested location (they don't see the exact coordinates, but which area was requested). On its own this is not very interesting data, but combined with other information it could also lead to localization or de-anonymization of targets.

There are potential other situations in which the IP address could leak, as this was never looked at with a holistic view, nor tested with wireshark. Automated tests would be useful to test against this problem in the future.

On Tails I suppose this problem would not appear, as Tails blocks every network traffic not going through the socks5 proxy, at least if the "unsafe browser" feature is not enabled. I did not test that though.

---

- Operating System (Linux/Mac/Windows/iOS/Android): Linux
- Delta Chat Version: 1.32.0
- Expected behavior: when using a socks5 proxy, all traffic is tunneled through the socks5 proxy
- Actual behavior: DNS and HTTP(S) traffic is sent over the normal network connection, without the proxy.
- Steps to reproduce the problem:
  - (capture traffic with wireshark or so)
  - Login to an account with Delta Chat
  - Enable location tracking
  - receive a location by someone else, look at it on the map
  - (check wireshark whether the traffic was sent over proxy or not)

**Comment by missytake** [comment]
Regarding 1., it seems that this DNS lookup is skipped if socks5 is enabled:

- https://github.com/deltachat/deltachat-core-rust/blob/4615c84f310ef53d2919e3de18cb8cf909e6d959/deltachat-ffi/src/lib.rs#L4202-4219 checks if socks5 is enabled and passes it to:
- https://github.com/deltachat/deltachat-core-rust/blob/master/src/provider.rs#L116, which skips the MX lookup.

So it seems to avoid this check the user needs to enable socks5 before typing the email address into the login form?

related discussion: https://github.com/deltachat/deltachat-desktop/pull/2355#issuecomment-917420191

**Comment by missytake** [comment]
Regarding 2.1, it seems that the autoconfig lookup is also skipped if socks5 is enabled:

- https://github.com/deltachat/deltachat-core-rust/blob/7a47c9e38bb8d25601b21765439edd0dba6c5518/src/configure.rs#L267-274 checks if socks5 is enabled and skips autoconfig if yes.

But as with the MX record lookup during configuration, if the user doesn't enable socks5 before typing the email address into the login form, the autoconfig HTTPS request happens.

**Comment by Simon-Laux** [comment]
is this still an issue in 1.34.4?

**Comment by farooqkz** [comment]
mapbox requests seem to happen in the desktop using `mapbox-gl`. I would suggest some solution to proxy ALL electron traffics through the SOCKS5 proxy. What do you think? @Simon-Laux  @link2xt  @missytake

**Comment by Simon-Laux** [comment]
core has a function to do http requests (`get_http_response` in jsonrpc api) that respects socks proxy settings.
The "real" solution would be to replace mapbox with leaflet and then use `get_http_response` from the jsonrpc to get the map tiles for leaflet, possible also cache them with pouchDB or sth like that.
Then we could block all electron triggered requests (we currently whitelist mapbox).

But replacing mapbox with leaflet is a bigger effort.

**Comment by farooqkz** [comment]
@Simon-Laux How replacing mapbox with leaflet helps here?

**Comment by Simon-Laux** [comment]
> @Simon-Laux How replacing mapbox with leaflet helps here?

less nodes dependencies is good in general and in leaflet we can easily make it fetch tiles over core. maybe map box too has too has an option for it too but we want to switch to leaflet anyway.

**Comment by r10s** [comment]
note, that there is a already a basic leaflet webxdc implementation on ios, it is not finished and not yet on master, however, i'd say we wait on desktop until that is done and consider moving away from mapbox then (maybe sth. for autumn) (i just explained things a little bit more at https://github.com/deltachat/deltachat-ios/pull/1912 )

**Comment by WofWca** [comment]
Other things to consider:
- [Video calls](https://github.com/deltachat/deltachat-desktop/blob/a5c02433976909fe01a64f3049c96d0d101ae9d8/packages/shared/constants.ts#L34-L35): we use third-party services, and just generate links to them.
- Various `runtime.openLink()` calls. E.g. "What is webxdc", ["Report bugs"](https://github.com/deltachat/deltachat-desktop/blob/1103f3581eb5588522cd813ef5477b1853fa7c9e/packages/frontend/src/components/dialogs/FullscreenMedia.tsx#L173), "Donate".

**Comment by nicodh** [comment]
> note, that there is a already a basic leaflet webxdc implementation on ios, it is not yet merged, however, i'd say we wait on desktop until that is done and consider moving away from mapbox then (maybe sth. for autumn) (i just explained things a little bit more at [deltachat/deltachat-ios#1912](https://github.com/deltachat/deltachat-ios/pull/1912) )

https://github.com/deltachat/deltachat-desktop/pull/5455 will use get_http_response for fetching the map tiles.

**Comment by nicodh** [comment]
> Other things to consider:
> 
> * [Video calls](https://github.com/deltachat/deltachat-desktop/blob/a5c02433976909fe01a64f3049c96d0d101ae9d8/packages/shared/constants.ts#L34-L35): we use third-party services, and just generate links to them.
> * Various `runtime.openLink()` calls. E.g. "What is webxdc", ["Report bugs"](https://github.com/deltachat/deltachat-desktop/blob/1103f3581eb5588522cd813ef5477b1853fa7c9e/packages/frontend/src/components/dialogs/FullscreenMedia.tsx#L173), "Donate".

Are these an issue? How could/should that be avoided?

**Comment by WofWca** [comment]
> Are these an issue? How could/should that be avoided?

Maybe an "Are you sure you want to open this in your browser?" dialog. Or otherwise making it clear that a web link will get opened.

**Comment by WofWca** [comment]
Meanwhile I think we are quite close to resolving this, or at least solidifying. Most of the web traffic is tunneled through core, `rpc.getHttpResponse`, after https://github.com/deltachat/deltachat-desktop/pull/5455 is merged. So now Electron's main window doesn't need to produce any traffic, and we can even set an invalid proxy for it (like we do for WebXDC windows on Tauri).

Things that are still remaining:
- Proxy the new calls' traffic: https://github.com/chatmail/core/issues/7309.
- Proxy WebXDC real-time channels traffic (Iroh).

**Comment by missytake** [comment]
webRTC Calls (and maybe webxdc real-time-channels?) are using UDP packets - [tor can not handle them](https://gitlab.torproject.org/legacy/trac/-/issues/7830), other proxies might be able to.

But if no UDP connection is possible via the proxy, what do we do? Fall back to not using the proxy, or show an error?

In https://github.com/chatmail/core/issues/7309 @r10s suggested that proxies are there to make it work at all, not to make it secure, then it would be about expectation management.
We could do this expectation management in blog posts like https://lefherz.net/2022/08/28/how-to-setup-delta-chat-on-tails-in-4-steps/. I don't know of any other calls apps disclosing that calls will not be proxied via tor, most just fall back to not using the proxy to make it work at all. Most messengers like Signal, Wire, or Element simply don't promote tor support, enabling tor for them is kinda hacky. Setting a proxy is typically a command line argument for them, not integrated into the app.

Personally, when the UDP connection doesn't work and a proxy is enabled, I'd like to show an error, and offer users to disable the proxy just for this call. But I have no idea if this technically feasible with the separation of layers etc., and how we even discover that UDP isn't supported by the proxy.

edit: uuuh tor is considering support UDP in arti, I'm excited: https://gitlab.torproject.org/tpo/core/arti/-/issues/463

**Comment by WofWca** [comment]
> webRTC Calls (and maybe webxdc real-time-channels?) are using UDP packets

WebRTC can also utilize TCP, in particular in these "proxy doesn't support UDP" situations. As a consequence, you _can_ do WebRTC calls over Tor. See <https://www.rfc-editor.org/rfc/rfc8828#name-modes-and-recommendations>, "Mode 4 - Force proxy":

> If the proxy does not support UDP (as is the case for all HTTP and most SOCKS [[RFC1928](https://www.rfc-editor.org/rfc/rfc8828#RFC1928)] proxies), or the WebRTC implementation does not support UDP proxying, the use of UDP will be disabled, and TCP will be used to send and receive media through the proxy. Use of TCP will result in reduced media quality, in addition to any performance considerations associated with sending all WebRTC media through the proxy server.

(note that this is only about the "Mode 4", which is not the default).

> proxies are there to make it work at all, not to make it secure

Yep, proxying calls traffic for the purpose of censorship circumvention also makes sense. If the Chatmail server is blocked for you, you cannot access it directly to do the call. So you have to use a proxy.

> Personally, when the UDP connection doesn't work and a proxy is enabled, I'd like to show an error, and offer users to disable the proxy just for this call.

This can be controlled, again, by setting the "IP handling policy". The default mode, mode 1, lets WebRTC use all possible interfaces, i.e. including direct connection. For privacy I think we should let users control the IP handling policy mode, e.g. with a toggle like "Allow calls to bypass VPN / proxy" (Signal has a similar toggle).
But again, WebRTC can utilize TCP.

**Comment by Simon-Laux** [comment]
With the removal of the option to turn off webxdc realtime channels (https://github.com/deltachat/deltachat-desktop/pull/5925), it will be interesting if iroh connections go through to tor or not, I guess they do not.
Either way would be interesting to check if iroh goes through system-wide vpn if there is any.

**Comment by r10s** [comment]
this issue is getting more into a discussion, and is hard to follow.

for [IP-addresses](https://delta.chat/en/help#who-sees-my-ip-address), current stance is to use a VPN, if that is a concern, and indeed, if i remember internal discussions correctly, proxy is more about enabling connectivity

i re-read, and it seems not actionable item is left, the original raised issues all have been targeted. closing therefore.
